add `organization_preferences` resource and data source

mdb opened the pull request:
This is ~a WIP PR to add address~ addresses issue #481 by introducing an `organization_preferences` data source and resource.

mdb committed:
WIP: add org preferences data source

TODO:

* how to uniquely ID the org preferences?
* add `resource_organization_preferences`

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
set ID on organization_preferences data source

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
stringify org preferences resource home_dashboard_uid

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add basic resource_organization_preferences

Signed-off-by: Mike Ball <mikedball@gmail.com>

julienduchesne commented on grafana/data_source_organization_preferences.go:
```diff
@@ -0,0 +1,82 @@
+* [HTTP API](https://grafana.com/docs/grafana/latest/developers/http_api/preferences/#get-current-org-prefs)
+`,
+		ReadContext: dataSourceOrganizationPreferencesRead,
+		Schema: map[string]*schema.Schema{
```
You can reuse your resource's schema with this function: https://github.com/grafana/terraform-provider-grafana/blob/master/grafana/schema.go#L23-L45

julienduchesne commented on grafana/data_source_organization_preferences.go:
```diff
@@ -0,0 +1,82 @@
+	// will typically not have the same datasource twice with the same parameters"
+	//
+	// However, in this instance, the data source does not accept any parameters; they are all computed.
+	// So, what would be a reasonable way to calculate its ID?
```
In that case, if every instance of that datasource has the same value, the id can be static. Just `organization_preferences` or something

julienduchesne commented on grafana/resource_organization_preferences.go:
```diff
@@ -0,0 +1,151 @@
+	d.Set("week_start", prefs.WeekStart)
+	d.Set("locale", prefs.Locale)
+
+	d.SetId(generateOrgPrefsIDSha())
```
Same as the datasource, the ID can be static, this resource is a singleton, we only expect to only ever be a single one

julienduchesne commented on grafana/resource_organization_preferences.go:
```diff
@@ -0,0 +1,151 @@
+func DeleteOrganizationPreferences(ctx context.Context, d *schema.ResourceData, meta interface{}) diag.Diagnostics {
+	client := meta.(*client).gapi
+	orgID, _ := strconv.ParseInt(d.Id(), 10, 64)
+	if err := client.DeleteOrg(orgID); err != nil {
```
I don't think deleting an org preferences resource should delete the org. It should just be a no-op IMO

mdb replied:
👍 I suspect this was early-stage copy/pasta. I've updated `DeleteOrganizationPreferences` to use [gapi.UpdateAllOrgPreferences](https://github.com/grafana/grafana-api-golang-client/blob/master/org_preferences.go#L37-L38) to clear out the org preferences via an empty `gapi.Preferences{}`. Does that seem reasonable?

julienduchesne commented on grafana/resource_organization_preferences.go:
```diff
@@ -0,0 +1,151 @@
+	}
+	if err := UpdateUsers(d, meta); err != nil {
+		return diag.FromErr(err)
+	}
```
I would've expected this function to be the exact same as the create one. This seems like it's the update function for the organization resource

mdb replied:
@julienduchesne I agree; I suspect this was early-stage copy pasta. I've updated `UpdateOrganizationPreferences` to simply return `CreateOrganizationPreferences`, given that both use [gapi.UpdateAllOrgPreferences](https://github.com/grafana/grafana-api-golang-client/blob/master/org_preferences.go#L38) under the hood. Is that reasonable?

julienduchesne reviewed (commented):
This would need some tests. Other than that and my comments, this looks nice!

mdb committed:
DRY up data_source_organization_preferences

This makes use of `cloneResourceSchemaForDatasource`
such that it's not necessary to redundantly declare
the schema across both the resource and data source.

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
'organization_preferences' as resource/data source ID

If every instance of a data source or resource has the same value,
the ID can be static, as that resource or data source is a
singleton.

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
'organization_preferences' as resource/data source ID

If every instance of a data source or resource has the same value,
the ID can be static, as that resource or data source is a
singleton.

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
clear org preferences on delete

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
properly update org preferences

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add org prefs 'timezone' validation

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add org prefs 'theme' validation

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
remove 'locale' from org prefs

This does not appear to be documented:
https://grafana.com/docs/grafana/latest/developers/http_api/preferences/

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add organization preferences docs

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
clarify org prefs available attr values

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
document organization_preferences datasource

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add initial organization_preferences datasource test

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add check for 'id' attr in organizaiton_preferences datasource

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add basic organization_preferences resource tests

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
enable organization_preferences tests on OSS

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
properly cast HomeDashboardID to int64

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
correct organization_preferences test assertions

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
test organization_preferences 'week_start' attr

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
add additional organization_preferences tests

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb committed:
test org preferences for Grafana versions >= 8

This seeks to avoid errors like:

```
=== RUN   TestAccResourceOrganizationPreferences
    resource_organization_preferences_test.go:28: Step 1/2 error: Check failed: Check 1/7 error: expected organization preferences week start 'Monday'; got ''
--- FAIL: TestAccResourceOrganizationPreferences (0.49s)
```

...as seen here:
https://drone.grafana.net/grafana/terraform-provider-grafana/1742/11/4

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented on grafana/resource_organization_preferences_test.go:
```diff
@@ -0,0 +1,120 @@
+
+func TestAccResourceOrganizationPreferences(t *testing.T) {
+	CheckOSSTestsEnabled(t)
+	CheckOSSTestsSemver(t, ">=8.0.0")
```
Is this `CheckOSSTestsSemver` acceptable? Without it, it seems that `WeekStart` is always `""` in at least Grafana 7.5.15, which results in acceptance test failures like...

```
=== RUN   TestAccResourceOrganizationPreferences
    resource_organization_preferences_test.go:14: Step 1/1 error: Check failed: Check 6/6 error: grafana_organization_preferences.test: Attribute 'week_start' expected "Tuesday", got ""
--- FAIL: TestAccResourceOrganizationPreferences (0.47s)
```

...as seen here: https://drone.grafana.net/grafana/terraform-provider-grafana/1741/11/4

julienduchesne replied:
Yes, that works! I think v7 is not even supported anymore

mdb committed:
add third org preferences resource.TestStep

Signed-off-by: Mike Ball <mikedball@gmail.com>

mdb commented:
> This would need some tests. Other than that and my comments, this looks nice!

@julienduchesne Would you mind offering another review, please? Are there any additional changes and/or tests you'd like to see?

julienduchesne reviewed (approved):
LGTM! Thanks for adding tests!
